### PR TITLE
STRF-6701 Reduce usage of Lodash

### PR DESCRIPTION
**Pull request by bc-williamkwon**
## What? Why?
 Refactored helper to use vanilla javascript methods instead.

https://jira.bigcommerce.com/browse/STRF-6701 

## How was it tested?

----
Unit tests still pass.
cc @bigcommerce/storefront-team

### CODE REVIEW

File: helpers/all.js [hunk 1]
```diff
@@ -21,20 +19,15 @@ const factory = () => {
...
         // Check if all the arguments are valid / truthy
-        result = _.all(args, function (arg) {
-            if (_.isArray(arg)) {
+        result = args.every(arg => {
+            if (Array.isArray(arg) === true) {
                 return !!arg.length;
-            }
-            // If an empty object is passed, arg is false
-            else if (_.isEmpty(arg) && _.isObject(arg)) {
+            } else if (arg && arg.constructor === Object && Object.keys(arg).length === 0) {
```
[the remark this answers is not in the document]
Comment (mattolson) [reply helpers/all.js]: I would also pull this and any other utility methods that are common throughout this PR into a library module.

File: helpers/any.js [hunk 2]
```diff
@@ -24,22 +22,24 @@ const factory = () => {
         opts = args.pop();
         predicate = opts.hash;
 
-        if (!_.isEmpty(predicate)) {
+        if (Object.keys(predicate).length !== 0) {
```
Comment (mattolson) [inline helpers/any.js]: `> 0` is probably clearer here

File: helpers/any.js [hunk 3]
```diff
@@ -24,22 +22,24 @@ const factory = () => {
...
             // With options hash, we check the contents of first argument
-            any = _.any(args[0], predicate);
+            any = args[0].some(arg => {
```
Comment (mattolson) [inline helpers/any.js]: `args[0]` could be either an object or an array, but we are assuming it's an array here. The lodash implementation supports both: https://github.com/lodash/lodash/blob/3.10.1/lodash.js#L7044

Comment (mattolson) [reply helpers/any.js]: Here's some documentation on this helper to help produce a correct implementation: https://developer.bigcommerce.com/stencil-docs/handlebars-syntax-and-helpers/handlebars-helpers-reference/new-subpage/loop-control-flow-helpers#handlebars_any

File: helpers/contains.js [hunk 4]
```diff
@@ -12,8 +10,14 @@ const _ = require('lodash');
 const factory = () => {
     return function() {
         var args = Array.prototype.slice.call(arguments, 0, -1),
```
Comment (mattolson) [inline helpers/contains.js]: We need to avoid slicing the arguments array.

File: helpers/deprecated/pick.js [hunk 5]
```diff
@@ -1,13 +1,18 @@
...
     /**
      * @deprecate
      */
-    return function() {
-        return _.pick.apply(null, arguments);
+    return function(obj, ...args) {
+        const pickedElements = {};
+
+        args.forEach(e => {
+            if (obj[e]) {
```
Comment (mattolson) [inline helpers/deprecated/pick.js]: You should probably use `hasOwnProperty` here in case the value is falsy.

File: helpers/getImage.js [hunk 6]
```diff
@@ -10,14 +8,19 @@ const factory = globals => {
         var size;
         var width;
         var height;
-
-        if (!_.isPlainObject(image) || !_.isString(image.data) || image.data.indexOf('{:size}') === -1) {
+        function isObject(obj) {
```
Comment (mattolson) [inline helpers/getImage.js]: Let's move these into a common library module

**Comment by mattolson** [comment]
Two references that can help guide this refactor:
* https://github.com/lodash/lodash/blob/3.10.1/lodash.js
* https://developer.bigcommerce.com/stencil-docs/handlebars-syntax-and-helpers/handlebars-helpers-reference/handlebars-full

File: helpers/all.js [hunk 7]
```diff
@@ -21,20 +19,15 @@ const factory = () => {
         opts = args.pop();
 
         // Check if all the arguments are valid / truthy
-        result = _.all(args, function (arg) {
-            if (_.isArray(arg)) {
+        result = args.every(arg => {
+            if (Array.isArray(arg) === true) {
```
[the remark this answers is not in the document]
Comment (bc-williamkwon) [reply helpers/all.js]: ok will change.

File: helpers/all.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (bc-williamkwon) [reply helpers/all.js]: ok will change.

File: helpers/any.js (hunk 2 above)
Comment (bc-williamkwon) [reply helpers/any.js]: ok will change.

File: helpers/any.js (hunk 3 above)
Comment (bc-williamkwon) [reply helpers/any.js]: ok will change.

File: helpers/getImage.js (hunk 6 above)
Comment (bc-williamkwon) [reply helpers/getImage.js]: ok will do.

File: helpers/deprecated/pick.js (hunk 5 above)
Comment (bc-williamkwon) [reply helpers/deprecated/pick.js]: ok will do.

File: helpers/contains.js (hunk 4 above)
Comment (bc-williamkwon) [reply helpers/contains.js]: ok.  will change it here too.

**Comment by mattolson** [comment]
Please rebase after your `arguments` PR which changed a lot of things.

File: helpers/lib/common.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+function isObject(val) {
+    return {}.toString.call(val) === '[object Object]';
+}
+function isOptions(val) {
```
Comment (mattolson) [inline helpers/lib/common.js]: Prefer extra line between functions

File: helpers/lib/common.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    return isObject(val) && val.fn;
+}
+function isString(val) {
+    return typeof val === 'string';
+}
+function isEmptyObject(val) {
+    return isObject(val) && Object.keys(val).length === 0;
+}
+module.exports = {isObject, isOptions, isString, isEmptyObject};
```
Comment (mattolson) [inline helpers/lib/common.js]: Prefer a multiline object here
```
module.exports = {
  isObject,
  isOptions,
  isString,
  isEmptyObject
};
```

File: helpers/lib/fonts.js [hunk 10]
```diff
@@ -18,34 +16,29 @@ const fontProviders = {
...
-            var collection = [],
-                familyHash = {};
+            const collection = [];
+            const familyHash = {};
 
-            _.each(fonts, function fontsIterator(font) {
-                var split = font.split('_'),
-                    familyKey = split[1],  // Eg: Open+Sans
-                    weights = split[2];    // Eg: 400,700italic
+            fonts.forEach(font => {
```
Comment (mattolson) [inline helpers/lib/fonts.js]: I think forEach is not going to give us a performance benefit. We should use a plain for loop: https://medium.com/tech-tajawal/loops-performances-in-node-js-9fbccf2d6aa6

Similarly for any other usages of forEach.

File: helpers/lib/fonts.js [hunk 11]
```diff
@@ -104,38 +96,41 @@ module.exports = function(format, themeSettings, handlebars, options) {
...
 
-        collectedFonts[provider].push(value);
+        collectedFonts[provider].push(themeSettings[setting]);
     });
 
     // Parse font strings based on provider
-    const parsedFonts = _.mapValues(collectedFonts, function(value, key) {
-        return fontProviders[key].parser(value);
-    });
+    const parsedFonts = Object.keys(collectedFonts).reduce((parsedFonts, font) => {
```
Comment (mattolson) [inline helpers/lib/fonts.js]: reduce, while nice and easy to read, is going to be slower than a for loop: https://hackernoon.com/3-javascript-performance-mistakes-you-should-stop-doing-ebf84b9de951

Comment (bc-williamkwon) [reply helpers/lib/fonts.js]: fixed.

File: helpers/lib/fonts.js (hunk 10 above)
Comment (bc-williamkwon) [reply helpers/lib/fonts.js]: fixed.

File: helpers/lib/common.js (hunk 9 above)
Comment (bc-williamkwon) [reply helpers/lib/common.js]: fixed.

File: helpers/lib/common.js (hunk 8 above)
Comment (bc-williamkwon) [reply helpers/lib/common.js]: fixed.

File: helpers/lib/common.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+function isString(val) {
+    return typeof val === 'string';
+}
+
+function isEmptyObject(val) {
+    return isObject(val) && Object.keys(val).length === 0;
+}
+
+function truthCheck(val) {
```
Comment (mattolson) [inline helpers/lib/common.js]: Let's call this `isTruthy` (that's kind of a semi-standard term for this)

File: helpers/all.js [hunk 13]
```diff
@@ -12,22 +10,16 @@ const factory = () => {
...
-                return !!arg.length;
-            }
-            // If an empty object is passed, arg is false
-            else if (_.isEmpty(arg) && _.isObject(arg)) {
-                return false;
+        for (let i = 0; i < args.length; i++) {
+            if (!common.truthCheck(args[i])) {
+                result = false;
+                break;
+            } else if (i === args.length - 1) {
```
Comment (mattolson) [inline helpers/all.js]: A more readable way to do this would be:
```
let result = true;
for (let i = 0; i < args.length; i++) {
  result = result && common.isTruthy(args[i]);
  if (!result) {
    break;
  }
}
```

File: helpers/any.js [hunk 14]
```diff
@@ -15,27 +14,10 @@ const factory = () => {
         const opts = args.pop();
         const predicate = opts.hash;
 
-        if (!_.isEmpty(predicate)) {
+        if (Object.keys(predicate).length > 0) {
```
Comment (mattolson) [inline helpers/any.js]: Let's use `!isEmptyObject` here.

File: helpers/any.js [hunk 15]
```diff
@@ -15,27 +14,10 @@ const factory = () => {
...
             // With options hash, we check the contents of first argument
             any = _.any(args[0], predicate);
-        } else {
```
Comment (mattolson) [inline helpers/any.js]: Not sure we want to remove this. I don't know that we've announced this deprecation anywhere.

File: helpers/lib/fonts.js [hunk 16]
```diff
@@ -18,35 +16,33 @@ const fontProviders = {
...
+                   continue;
                 }
-
-                if (_.isUndefined(weights)) {
+                if (typeof weights === 'undefined') {
                     weights = '';
                 }
-
-                if (!_.isArray(familyHash[familyKey])) {
+                if (Array.isArray(familyHash[familyKey]) === false) {
```
Comment (mattolson) [inline helpers/lib/fonts.js]: Should just do `!Array.isArray`

File: helpers/lib/fonts.js [hunk 17]
```diff
@@ -18,35 +16,33 @@ const fontProviders = {
...
                     familyHash[familyKey] = [];
                 }
-
                 weights = weights.split(',');
-
                 familyHash[familyKey].push(weights);
-                familyHash[familyKey] = _.uniq(_.flatten(familyHash[familyKey]));
-            });
+                familyHash[familyKey] = [...new Set([...familyHash[familyKey]])];
```
Comment (mattolson) [inline helpers/lib/fonts.js]: Let's just use `Array.from(set)`

File: helpers/lib/fonts.js [hunk 18]
```diff
@@ -84,58 +86,76 @@ const fontProviders = {
...
  */
 module.exports = function(format, themeSettings, handlebars, options) {
-    
+
     const collectedFonts = {};
-    _.each(themeSettings, function(value, key) {
-        //check that -font is on end of string but not start of string
+    const themeSettingsKeys = Object.keys(themeSettings);
+    for (let i = 0; i < themeSettingsKeys.length; i++) {
+        const setting = themeSettingsKeys[i];
```
Comment (mattolson) [inline helpers/lib/fonts.js]: I would put
```
const key = themeSettingsKeys[i];
const value = themeSettings[key];
```
at the top of the loop and that will minimize changes below and probably make things easier to read

File: helpers/limit.js [hunk 19]
```diff
@@ -10,10 +8,10 @@ const _ = require('lodash');
  */
 const factory = () => {
     return function(data, limit) {
-        if (_.isString(data)) {
+        if (typeof data === 'string') {
```
Comment (mattolson) [inline helpers/limit.js]: Let's use `common.isString`

File: spec/helpers/any.js [hunk 20]
```diff
@@ -48,57 +48,3 @@ describe('any helper (with option hash)', function() {
     });
 });
 
-
-describe('any helper (with multiple arguments)', function() {
```
Comment (mattolson) [inline spec/helpers/any.js]: Should probably not remove this

File: helpers/or.js [hunk 21]
```diff
@@ -12,29 +11,22 @@ const factory = () => {
...
-        // Without options hash, we check all the arguments
-        const any = _.any(args, function (arg) {
-            if (_.isArray(arg)) {
-                return !!arg.length;
-            }
-            // If an empty object is passed, arg is false
-            else if (_.isEmpty(arg) && _.isObject(arg)) {
-                return false;
+        for (let i = 0; i < args.length; i++) {
+            if (common.truthCheck(args[i])) {
```
Comment (mattolson) [inline helpers/or.js]: Similar to the above, perhaps

```
let any = false;
for (let i = 0; i < args.length; i++) {
  any = any || common.isTruthy(args[i]);
  if (any) {
    break;
  }
}
```

File: helpers/pluck.js [hunk 22]
```diff
@@ -1,10 +1,21 @@
...
-        return _.pluck(collection, path);
+        const arr = [];
+        if (Array.isArray(collection)) {
+            for (let i = 0; i < collection.length; i++) {
+                arr.push(collection[i][path]);
+            }
+        } else if (common.isObject(collection)) {
+            const keys = Object.keys(collection);
+            for (let i = 0; i < keys.length; i++) {
+                arr.push(collection[keys[i]][path]);
```
Comment (mattolson) [inline helpers/pluck.js]: Don't we need some kind of existence check here? Otherwise, we'll push `undefined` into the array.

File: helpers/resourceHints.js [hunk 23]
```diff
@@ -16,28 +15,28 @@ const factory = globals => {
...
         // Add font providers
-        const fontProviders = _.keys(getFonts('providerLists', globals.getThemeSettings(), globals.handlebars));
-        _.each(fontProviders, function(provider) {
-            if (typeof fontResources[provider] !== 'undefined') {
-                hosts = hosts.concat(fontResources[provider]);
+        const providers = Object.keys(getFonts('providerLists', globals.getThemeSettings(), globals.handlebars));
+
+        for (let i = 0; i < providers.length; i++) {
+            if (typeof fontResources[providers[i]] !== 'undefined') {
+                hosts.push(...fontResources[providers[i]]);
```
Comment (mattolson) [inline helpers/resourceHints.js]: Why not `hosts.concat`?

File: index.js [hunk 24]
```diff
@@ -66,9 +64,10 @@ class HandlebarsRenderer {
         };
 
         // Register helpers with Handlebars
-        helpers.forEach(spec => {
+        for (let i = 0; i < helpers.length; i++) {
+            const spec = helpers[i];
             this.handlebars.registerHelper(spec.name, spec.factory(this.helperContext));
-        });
+        }
```
Comment (mattolson) [inline index.js]: Extra whitespace

File: helpers/resourceHints.js (hunk 23 above)
Comment (bc-williamkwon) [reply helpers/resourceHints.js]: Mainly to make hosts a const.

Comment (bc-williamkwon) [reply helpers/resourceHints.js]: It returns a new array too so I think it's doing more work.

File: helpers/lib/common.js (hunk 12 above)
Comment (bc-williamkwon) [reply helpers/lib/common.js]: done.

File: helpers/all.js (hunk 13 above)
Comment (bc-williamkwon) [reply helpers/all.js]: done.

File: helpers/any.js (hunk 14 above)
Comment (bc-williamkwon) [reply helpers/any.js]: done.

File: helpers/lib/fonts.js (hunk 16 above)
Comment (bc-williamkwon) [reply helpers/lib/fonts.js]: done.

File: helpers/lib/fonts.js (hunk 17 above)
Comment (bc-williamkwon) [reply helpers/lib/fonts.js]: done.

File: helpers/lib/fonts.js (hunk 18 above)
Comment (bc-williamkwon) [reply helpers/lib/fonts.js]: done.

File: helpers/limit.js (hunk 19 above)
Comment (bc-williamkwon) [reply helpers/limit.js]: done.

File: helpers/or.js (hunk 21 above)
Comment (bc-williamkwon) [reply helpers/or.js]: done.

File: index.js (hunk 24 above)
Comment (bc-williamkwon) [reply index.js]: done.

File: helpers/pluck.js (hunk 22 above)
Comment (bc-williamkwon) [reply helpers/pluck.js]: Based on documentation.  I thought it would be ok to not do that.  But I have changed it.

File: helpers/any.js (hunk 15 above)
Comment (bc-williamkwon) [reply helpers/any.js]: Unremoved.

File: spec/helpers/any.js (hunk 20 above)
Comment (bc-williamkwon) [reply spec/helpers/any.js]: unremoved.

File: helpers/lib/fonts.js [hunk 25]
```diff
@@ -84,58 +86,77 @@ const fontProviders = {
  * @returns {Object.<string, Array>|string}
  */
 module.exports = function(format, themeSettings, handlebars, options) {
-    
+
     const collectedFonts = {};
-    _.each(themeSettings, function(value, key) {
-        //check that -font is on end of string but not start of string
+    const themeSettingsKeys = Object.keys(themeSettings);
```
Comment (mattolson) [inline helpers/lib/fonts.js]: This portion of the change has become really complex, and I'm pretty sure something is wrong with this implementation. Perhaps we should strip this part out and save it for another PR?

File: helpers/pluck.js [hunk 26]
```diff
@@ -1,10 +1,25 @@
 'use strict';
-
-var _ = require('lodash');
+const common = require('./lib/common.js');
 
 const factory = () => {
     return function(collection, path) {
-        return _.pluck(collection, path);
+        const arr = [];
```
Comment (mattolson) [inline helpers/pluck.js]: What is the current behavior of this function? Should it return an object when passed an object, or is it supposed to always return an array?

File: helpers/resourceHints.js [hunk 27]
```diff
@@ -16,28 +15,28 @@ const factory = globals => {
...
             }
-        });
+        }
+
+        for (let j = 0; j < hosts.length; j++) {
```
Comment (mattolson) [inline helpers/resourceHints.js]: Perhaps instead of a second loop, we should just do this as part of the first loop?

File: index.js [hunk 28]
```diff
@@ -166,22 +165,20 @@ class HandlebarsRenderer {
      * @param {Object} A set of templates to register with handlebars
      */
     addTemplates(templates) {
-        _.each(templates, (template, path) => {
-            // Don't do this work twice, first one wins.
-            if (typeof this.handlebars.partials[path] !== 'undefined') {
+        const keys = Object.keys(templates);
+
+        for (let i = 0; i < keys.length; i++) {
+            const template = keys[i];
```
Comment (mattolson) [inline index.js]: This should be named `path`

File: index.js [hunk 29]
```diff
@@ -166,22 +165,20 @@ class HandlebarsRenderer {
      * @param {Object} A set of templates to register with handlebars
      */
     addTemplates(templates) {
-        _.each(templates, (template, path) => {
-            // Don't do this work twice, first one wins.
-            if (typeof this.handlebars.partials[path] !== 'undefined') {
+        const keys = Object.keys(templates);
```
Comment (mattolson) [inline index.js]: This should be named `paths`

File: index.js [hunk 30]
```diff
@@ -166,22 +165,20 @@ class HandlebarsRenderer {
...
+        for (let i = 0; i < keys.length; i++) {
+            const template = keys[i];
+            if (typeof this.handlebars.partials[template] !== 'undefined') {
                 return;
             }
-
-            // Check if it is a precompiled template
             try {
-                template = this._tryRestoringPrecompiled(template);
+                templates[template] = this._tryRestoringPrecompiled(templates[template]);
```
Comment (mattolson) [inline index.js]: Let's not mutate the incoming parameter. `template` should be a local variable declared inside the loop.

File: index.js [hunk 31]
```diff
@@ -221,14 +218,17 @@ class HandlebarsRenderer {
      */
     getPreProcessor() {
         return templates => {
+            const keys = Object.keys(templates);
```
Comment (mattolson) [inline index.js]: Let's name this `paths`

File: helpers/resourceHints.js (hunk 27 above)
Comment (bc-williamkwon) [reply helpers/resourceHints.js]: We don't have individual access to each host in the first.

File: helpers/pluck.js (hunk 26 above)
Comment (bc-williamkwon) [reply helpers/pluck.js]: .pluck is now map in lodash and returns a mapped array.  I used both _.pluck and my own against test cases and they pass.

Comment (bc-williamkwon) [reply helpers/pluck.js]: I think it would be better to keep lodash for collection helpers.

File: index.js (hunk 28 above)
Comment (bc-williamkwon) [reply index.js]: fixed.

File: index.js (hunk 29 above)
Comment (bc-williamkwon) [reply index.js]: fixed.

File: index.js (hunk 31 above)
Comment (bc-williamkwon) [reply index.js]: done.

File: index.js (hunk 30 above)
Comment (bc-williamkwon) [reply index.js]: fixed.

File: index.js [hunk 32]
```diff
@@ -166,22 +165,21 @@ class HandlebarsRenderer {
      * @param {Object} A set of templates to register with handlebars
      */
     addTemplates(templates) {
-        _.each(templates, (template, path) => {
-            // Don't do this work twice, first one wins.
+        const paths = Object.keys(templates);
+
+        for (let i = 0; i < paths.length; i++) {
+            const path = paths[i];
+            let template;
```
Comment (mattolson) [inline index.js]: I would move this closer to its usage, i.e. right below the comment `// Check if it is a precompiled template`

File: helpers.js [hunk 33]
```diff
@@ -3,21 +3,23 @@
...
 
 // Load helpers
-fs.readdirSync(Path.join(__dirname, 'helpers')).forEach(filename => {
-    if (!fs.lstatSync(Path.join(__dirname, 'helpers', filename)).isDirectory()) {
-        helpers = helpers.concat(require('./helpers/' + filename));
-    }
-});
+const notDeprecated = fs.readdirSync(Path.join(__dirname, 'helpers'));
 
+for (let i = 0; i < notDeprecated.length; i++) {
```
Comment (mattolson) [inline helpers.js]: Not important to refactor this file. This only gets loaded once, and the incremental gain of using a plain for loop is negligible compared to loading files from the file system. Let's revert the changes to this file.

File: helpers/all.js [hunk 34]
```diff
@@ -12,22 +10,14 @@ const factory = () => {
     return function(...args) {
         // Take the last arg (which is a Handlebars options object) out of args array
         const opts = args.pop();
-
+        let result;
```
Comment (mattolson) [inline helpers/all.js]: Minor nitpick, but I would put this below the comment

File: helpers/lib/fonts.js [hunk 35]
```diff
@@ -18,35 +18,33 @@ const fontProviders = {
...
+                if (!Array.isArray(familyHash[familyKey])) {
                     familyHash[familyKey] = [];
                 }
-
                 weights = weights.split(',');
-
                 familyHash[familyKey].push(weights);
-                familyHash[familyKey] = _.uniq(_.flatten(familyHash[familyKey]));
-            });
+                familyHash[familyKey] = Array.from(new Set([...familyHash[familyKey]]));
```
Comment (mattolson) [inline helpers/lib/fonts.js]: This could probably be simplified

Instead of
```
familyHash[familyKey].push(weights);
familyHash[familyKey] = Array.from(new Set([...familyHash[familyKey]]));
```

how about
```
familyHash[familyKey].push(...weights);
```

and then in the final loop
```
for (let i = 0; i < familyHashKeys.length; i++) {
  const family = familyHashKeys[i];
  const weights = familyHash[family];
  const uniqueWeights = Array.from(new Set(weights));
  collection.push(`${family}:${uniqueWeights.join(',')}`);
}

```

File: index.js [hunk 36]
```diff
@@ -166,22 +165,21 @@ class HandlebarsRenderer {
...
             if (typeof this.handlebars.partials[path] !== 'undefined') {
                 return;
             }
-
-            // Check if it is a precompiled template
```
Comment (mattolson) [inline index.js]: Please don't remove this comment

File: index.js [hunk 37]
```diff
@@ -166,22 +165,21 @@ class HandlebarsRenderer {
...
             try {
-                template = this._tryRestoringPrecompiled(template);
+                template = this._tryRestoringPrecompiled(templates[path]);
             } catch(e) {
                 throw new FormatError(e.message);
             }
-
-            // Register it with handlebars
```
Comment (mattolson) [inline index.js]: Please don't remove this comment

File: index.js (hunk 32 above)
Comment (bc-williamkwon) [reply index.js]: fixed

File: helpers.js (hunk 33 above)
Comment (bc-williamkwon) [reply helpers.js]: fixed.

File: helpers/all.js (hunk 34 above)
Comment (bc-williamkwon) [reply helpers/all.js]: fixed.

File: helpers/lib/fonts.js (hunk 35 above)
Comment (bc-williamkwon) [reply helpers/lib/fonts.js]: fixed.

File: index.js (hunk 36 above)
Comment (bc-williamkwon) [reply index.js]: fixed.

File: index.js (hunk 37 above)
Comment (bc-williamkwon) [reply index.js]: fixed.

File: helpers/all.js (hunk 34 above)
Comment (bc-williamkwon) [reply helpers/all.js]: fixed.

File: helpers/lib/fonts.js (hunk 35 above)
Comment (bc-williamkwon) [reply helpers/lib/fonts.js]: fixed.

**Comment by bc-williamkwon** [comment]
Removed lodash methods from any, contains, and pluck.

File: helpers/lib/common.js [hunk 38]
```diff
@@ -0,0 +1,38 @@
...
+    if (Array.isArray(val)) {
+        return !!val.length;
+    } else if (isEmptyObject(val)) {
+        return false;
+    } else {
+        return !!val;
+    }
+}
+
+function sameValueZero(val1, val2) {
```
Comment (mattolson) [inline helpers/lib/common.js]: I don't understand what this is trying to do. We probably need a comment here and probably a function name change.

Comment (bc-williamkwon) [reply helpers/lib/common.js]: sameValueZero is a Lodash algorithm that also allows for NaN === NaN to return true.

Comment (bc-williamkwon) [reply helpers/lib/common.js]: changed.

File: helpers/lib/fonts.js [hunk 39]
```diff
@@ -18,35 +16,35 @@ const fontProviders = {
...
-            });
+            const familyHashKeys = Object.keys(familyHash);
 
-            _.each(familyHash, function fontHashIterator(weights, family) {
-                collection.push(family + ':' + weights.join(','));
-            });
+            for (let i = 0; i < familyHashKeys.length; i++) {
+                const family = familyHashKeys[i];
+                const weights = familyHash[family];
+                const uniqueWeights = Array.from(new Set(weights));
```
Comment (bc-zoharmuzafi) [inline helpers/lib/fonts.js]: @bc-williamkwon you can just set in weights the unique values instead of defining a new var

File: helpers/lib/fonts.js [hunk 40]
```diff
@@ -84,54 +88,76 @@ const fontProviders = {
...
-            return;
+            continue;
         }
 
         const splits = value.split('_');
         const provider = splits[0];
 
         if (typeof fontProviders[provider] === 'undefined') {
-            return;
+            continue;
```
Comment (bc-zoharmuzafi) [inline helpers/lib/fonts.js]: Are we ok to change return to continue here?

File: helpers/resourceHints.js [hunk 41]
```diff
@@ -16,28 +15,25 @@ const factory = globals => {
...
+        }
+        return new globals.handlebars.SafeString(hosts.join(''));
     };
 };
 
 module.exports = [{
     name: 'resourceHints',
     factory: factory,
-}];
+}];
```
Comment (bc-zoharmuzafi) [inline helpers/resourceHints.js]: new line

**Comment by bc-zoharmuzafi** [comment]
@bc-williamkwon Overall the changes looks good. My concern is that this pr is changing very simple implementation and covers a lot of method. my suggestion due to the high risk, is to break this pr to really small prs, which cover one method each and to be sure each changed method backed up with sufficient unitest

File: helpers/lib/fonts.js (hunk 40 above)
Comment (bc-williamkwon) [reply helpers/lib/fonts.js]: yes, because return in a forEach is equivalent to continue in a for
